frontend/src/abilities.ts

```diff
@@ -1,53 +1,77 @@
 import { AbilityBuilder } from "@casl/ability";
 import { store } from "@/store";
 
-const TESubjects = [
-  "TE/Task",
-  "TE/Upload",
-  "TE/Allot",
-  "TE/Tasks",
-  "TE/MyTasks"
-];
-const SQRSubjects = ["SQR/Allot", "SQR/UpdateDone"];
-
-const subjects = [...TESubjects, ...SQRSubjects].reduce(
-  (permissions, permission) => {
-    permissions[permission] = permission;
-    return permissions;
+export const subjects = {
+  TE: {
+    tasks: "TE/Tasks",
+    myTasks: "TE/MyTasks",
+    task: "TE/Task"
   },
-  {} as { [key: string]: string }
-);
+  SQR: {
+    tasks: "SQR/Tasks",
+    form: "SQR/Form"
+  }
+};
+
+class Data {
+  constructor(props: any) {
+    Object.assign(this, props);
+  }
+}
+
+export class SQRForm extends Data {
+  static get modelName() {
+    return subjects.SQR.form;
+  }
+}
+
+export const SubjectsPlugin = {
+  install(Vue: any) {
+    Vue.prototype.$subjects = subjects;
+  }
+};
 
 const hasRole = (roles: any) => {
   return store.getters["user/hasRole"](roles);
 };
 
-const handleTEPermissions = (can: any) => {
+const handleTEPermissions = (can: any, cannot: any) => {
   if (hasRole("TE.coordinator")) {
-    can("manage", TESubjects);
-    return;
+    can("manage", subjects.TE.task);
+    can("manage", subjects.TE.tasks);
   }
   if (hasRole("TE.checker")) {
-    can("resolve", subjects["TE/Task"]);
+    can(["resolve", "view"], subjects.TE.task);
+    can("view", subjects.TE.tasks);
   }
   if (hasRole("TE.editor")) {
-    can("upload", subjects["TE/Upload"]);
+    can(["upload", "view"], subjects.TE.task);
+    can("view", subjects.TE.myTasks);
   }
 };
 
-const handleSQRPermissions = (can: any) => {
-  if (hasRole("SQR.coordinator")) {
-    can("manage", SQRSubjects);
+const handleSQRPermissions = (can: any, cannot: any) => {
+  can("save", subjects.SQR.form, { isCompleted: false });
+
+  if (!hasRole(["SQR.coordinator", "SQR.checker"])) {
+    can("submit", subjects.SQR.form, { isCompleted: false });
```

Comment:
The condition is if they are not `SQR.coordinator` or `SQR.checker`

Reply:
OK, I see.

Reply:
I feel it might be more explicit if all would be granted permission for `submit` action with `isCompleted: false` and elevated roles would be granted permission with `isCompleted: true`.